HIVE-25871: Hive should set name mapping table property for migrated …

boroknagyz opened the pull request:
…Iceberg tables

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
With this PR, Hive will set table property 'schema.name-mapping.default' for migrated Iceberg tables.
The value of this property contains a mapping between Iceberg field ids and column names.

### Why are the changes needed?
This table property is useful for column projection of legacy data files:
https://iceberg.apache.org/#spec/#column-projection

### Does this PR introduce _any_ user-facing change?
When users migrate a legacy table to Iceberg, this new table property will be set.

### How was this patch tested?
Extended unit test TestHiveIcebergMigration.
Q tests.

pvary reviewed (approved):
+1 pending tests

marton-bod commented on iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergMigration.java:
```diff
@@ -225,12 +226,15 @@ private void validateSd(Table hmsTable, String format) {
     } else {
       Assert.assertNull(migratedProp);
       Assert.assertNotEquals(BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE.toUpperCase(), tableTypeProp);
+      Assert.assertTrue(nameMappingProp == null);
```
nit: could use Assert.assertNull here

marton-bod reviewed (approved):
+1 pending tests

boroknagyz commented:
HiveSchemaConverter assigned field ids starting from 0, while Iceberg assigns field ids starting from 1.
This caused test failures because the name mapping had wrong field ids.
So in the second commit I fixed HiveSchemaConverter as well.

pvary commented:
> HiveSchemaConverter assigned field ids starting from 0, while Iceberg assigns field ids starting from 1. This caused test failures because the name mapping had wrong field ids. So in the second commit I fixed HiveSchemaConverter as well.

If we provide a schema where the id is starting from 0, would that mean that the schema is not correct? Or this is just to be more aligned with the way Iceberg generates ids?

boroknagyz commented:
> > HiveSchemaConverter assigned field ids starting from 0, while Iceberg assigns field ids starting from 1. This caused test failures because the name mapping had wrong field ids. So in the second commit I fixed HiveSchemaConverter as well.
> 
> If we provide a schema where the id is starting from 0, would that mean that the schema is not correct? Or this is just to be more aligned with the way Iceberg generates ids?

Seems like Iceberg doesn't respect the field ids during table creation, it always reassigns them.
E.g. I was trying to create a table with this schema:

`iceberg.mr.table.schema={"type":"struct","schema-id":0,"fields":[{"id":0,"name":"a","required":false,"type":"int"}]}`

or

`iceberg.mr.table.schema={"type":"struct","schema-id":0,"fields":[{"id":5,"name":"a","required":false,"type":"int"}]}`

And Iceberg created the following schema:

`table {
  1: a: optional int
}`

OTOH, I don't think the value 0 is invalid. The spec reserves only these ids:
https://iceberg.apache.org/#spec/#reserved-field-ids

So I edited the Iceberg snapshot file and rewrote the field ids to start from 0, then inserted data files and queried the table. Iceberg didn't complain about the 0 field id.

boroknagyz commented on iceberg/iceberg-handler/src/test/queries/positive/describe_iceberg_table.q:
```diff
@@ -8,7 +8,7 @@ DROP TABLE IF EXISTS ice_t_transform;
 
 DROP TABLE IF EXISTS ice_t_transform_prop;
-CREATE EXTERNAL TABLE ice_t_transform_prop (id int, year_field date, month_field date, day_field date, hour_field timestamp, truncate_field string, bucket_field int, identity_field int) STORED BY ICEBERG TBLPROPERTIES ('iceberg.mr.table.partition.spec'='{"spec-id":0,"fields":[{"name":"year_field_year","transform":"year","source-id":1,"field-id":1000},{"name":"month_field_month","transform":"month ...
+CREATE EXTERNAL TABLE ice_t_transform_prop (id int, year_field date, month_field date, day_field date, hour_field timestamp, truncate_field string, bucket_field int, identity_field int) STORED BY ICEBERG TBLPROPERTIES ('iceberg.mr.table.partition.spec'='{"spec-id":0,"fields":[{"name":"year_field_year","transform":"year","source-id":2,"field-id":1000},{"name":"month_field_month","transform":"month ...
```
I've issued these commands in Hive (with the original 0-based field ids), then checked the Iceberg snapshot file and the field ids were transformed by Iceberg to 1-based anyway. So now the values in the statements are the same that the values in the snapshots.

marton-bod replied:
I'm probably missing something obvious - can you explain why the source-id values had to be incremented? Thanks!

boroknagyz replied:
Prior to this patch `HiveSchemaConverter` used 0-based indexing when it assigned the field ids. E.g. in the above statement it would assign field id 0 to `id`, field id 1 to `year_field`, and so on. Hence in 'iceberg.mr.table.partition.spec' the source-id 1 referred to the `year_field`. Everything was fine, but when Iceberg creates a table it reassigns the field ids using 1-based indexing (field id 1 is `id`, field id 2 is `year_field`). And Iceberg is smart enough to use the correct ids in the partition spec, i.e. it replaces source id 1 to source id 2 and so on.

So everything worked OK, but you had to specify different field/source ids in Hive than the actual field/source ids assigned by Iceberg.

With this change, you need to use the same 1-based indexing in the partition spec that Iceberg will use later.

marton-bod replied:
Makes sense, thx!

marton-bod commented:
LGTM, will merge this today unless @pvary has further comments

pvary commented:
LGTM +1.
I think this change should go into the Iceberg repo as well. What do you think?

marton-bod commented:
Thanks for the contribution @boroknagyz! As Peter mentioned it would be great to get the relevant parts into the upstream Iceberg code base as well - is this something you would be fancy doing too?

boroknagyz commented:
Ah right, things are currently being duplicated between Hive and Iceberg. Sure, I'll happily add these changes to Iceberg as well!

boroknagyz commented:
FYI I've uploaded a PR to Iceberg: https://github.com/apache/iceberg/pull/3947

It only contains the 1-based indexing of this PR as the table migration code is not present in the Iceberg repo.
